core/src/main/java/org/infinispan/executors/ManageableExecutorService.java

```diff
@@ -0,0 +1,87 @@
+package org.infinispan.executors;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.jmx.annotations.DataType;
+import org.infinispan.jmx.annotations.DisplayType;
+import org.infinispan.jmx.annotations.MBean;
+import org.infinispan.jmx.annotations.ManagedAttribute;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 10.0
+ **/
+@MBean
+@Scope(Scopes.GLOBAL)
+public abstract class ManageableExecutorService<T extends ExecutorService> {
+
+   protected T executor;
+
+   @ManagedAttribute(
+         description = "Returns the number of threads in this executor.",
+         displayName = "Number of executor threads",
+         dataType = DataType.TRAIT,
+         displayType = DisplayType.SUMMARY
+   )
+   int getPoolSize() {
+      if (executor instanceof ThreadPoolExecutor) {
+         return ((ThreadPoolExecutor) executor).getPoolSize();
+      } else {
+         return -1;
+      }
+   }
+
+   @ManagedAttribute(
+         description = "Returns the number of active executor threads.",
+         displayName = "Number of active executor threads",
+         dataType = DataType.TRAIT,
+         displayType = DisplayType.SUMMARY
+   )
+   int getActiveCount() {
+      if (executor instanceof ThreadPoolExecutor) {
+         return ((ThreadPoolExecutor) executor).getActiveCount();
+      } else {
+         return -1;
+      }
+   }
+
+   @ManagedAttribute(
+         description = "Returns the maximum number of executor threads.",
+         displayName = "Maximum number of executor threads",
+         dataType = DataType.TRAIT,
+         displayType = DisplayType.SUMMARY,
+         writable = true
+   )
+   int getMaximumPoolSize() {
+      if (executor instanceof ThreadPoolExecutor) {
+         return ((ThreadPoolExecutor) executor).getMaximumPoolSize();
+      } else {
+         return -1;
+      }
+   }
+
+   void setMaximumPoolSize(int maximumPoolSize) {
+      if (executor instanceof ThreadPoolExecutor) {
+         ((ThreadPoolExecutor) executor).setMaximumPoolSize(maximumPoolSize);
```

Comment:
For executors that have a queue, we set the core size to be the same as the maximum size, so setting the maximum size should also change the core size.

Maybe we also need attributes for the queue size and the maximum queue size.

Reply:
Ok for the core size.
As for the queue, I'll add getQueueSize() and getQueueRemainingCapacity() which map to the underlying equivalent methods in the queue impl.

Reply:
Unfortunately it's still not correct, the core size is 0 for executors that don't have a queue.

So we only need to update the core size if a queue is present, and we should probably add a comment about how bad ThreadPoolExecutor is at creating new threads when a queue is present. Maybe even a link to https://issues.jboss.org/browse/ISPN-4585, although it needs a more exotic example now that prepare commands don't block threads (e.g. state transfer, which still blocks a lot and has commands depending on each other).